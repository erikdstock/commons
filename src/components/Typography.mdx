---
Typography
---
import { Text, H1, H2, H3, H4 } from './Typography'
import { Box, Flex } from 'rebass'
import { Playground } from 'docz'
import { lorem } from '../docz/support'


# Typography

## Headings

  <Box>
    <H1>Heading 1</H1>
    <H2>Heading 2</H2>
    <H3>Heading 3</H3>
    <H4>Heading 4</H4>
  </Box>

## Font Faces

  <H3 fontFamily="sans">
    Sans
  </H3>
  <Text fontFamily="sans">
    New York City Democratic Socialists is the NYC local of Democratic Socialists of America,
    the largest socialist organization in the United States with about 30,000 members (as of November 2017).
    NYC-DSA is run by its 3,000+ members and activists who are working together to build a democratic socialist
    organization in the five boroughs. We are socialists because we share a vision of a humane social order based on
    popular control of resources and production, economic planning, equitable distribution, feminism,
    racial equality, and non-oppressive relationships. 
  </Text>
  <H3 fontFamily="serif">
    Serif
  </H3>
  <Text fontFamily="serif">
    We are socialists because we are opposed to an
    economy organized for private profit that produces gross inequalities of wealth and power, discrimination
    based on race and sex, and brutality and violence in defense of the status quo. We are socialists because
    we want to work together to develop a concrete strategy for achieving that vision, for achieving a transition
    to democratic socialism in America. 
  </Text>
  <H3 fontFamily="mono">
    Monospace
  </H3>
  <Text fontFamily="mono">
    We believe that such a strategy must acknowledge the class structure of
    American society and that this class structure means that there is a basic conflict of interest between
    those sectors with enormous economic power and the vast majority of the population.
    And we believe that only democratic movements and socialist parties can lead the way to a better world.
  </Text>



