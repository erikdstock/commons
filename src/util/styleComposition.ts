import styled from "@emotion/styled"
import ssCss, { CSSObject, SystemStyleObject } from "@styled-system/css"

import get from "lodash/get"
import { RosesTheme } from "../theme"

/** Clumsily force styledCss function to believe that there will be a theme present */
type StyledCssFn = (o: SystemStyleObject) => CSSObject
const styledCss = ssCss as StyledCssFn

/** The base style for Box and everything on up. */
export const boxStyle: CSSObject = {
  boxSizing: "border-box",
}

interface ComposeOptions {
  defaultVariant?: string
}

const themeVariantsRoot: keyof RosesThemeObject = "variants"
const componentStylesRoot: keyof RosesThemeObject = "componentStyles"

interface StyledInterpolationProps extends RosesStyleProps {
  theme: RosesThemeObject
}

type InterpolationFn = (p: StyledInterpolationProps) => CSSObject | undefined

/** Reach into the theme, get a style object, apply it with styledCss */
const themed: (path: string) => InterpolationFn = path => props => {
  console.warn(path, props.theme)
  const componentStyle = get(props.theme, path)
  console.warn("Applying Style: ", componentStyle)
  return componentStyle && styledCss(componentStyle)
}

const themedComponent: (
  name: string
) => (p: StyledInterpolationProps) => CSSObject | undefined = (
  name: string
) => {
  console.warn("wrapping component with " + name + " styles")
  return themed(`${componentStylesRoot}.${name}`)
}

const rxHandler: InterpolationFn = ({ rx }) => rx && styledCss(rx)

type VariantHandler = (key: string, op: ComposeOptions) => InterpolationFn

const variantHandler: VariantHandler = (componentKey, ops) => ({
  theme,
  variant,
}) => {
  const selectedVariant = variant || ops.defaultVariant
  if (selectedVariant) {
    const base = `${themeVariantsRoot}.${componentKey}`
    const variantCss = get(theme, `${base}.${selectedVariant}`)
    console.warn("variant: ", variantCss)
    return variantCss && styledCss(variantCss)
  }
}

/**
 * Wrap a base styled component with:
 *   - a responsive, theme-aware style `css` prop
 *   - a `variant` prop for accessing styles set on the theme.variants key
 * @example:
 * ```tsx
 *  const RawBox = styled('div')({boxSizing: 'border-box'})
 *  const Widget = withStyleProps('Box', RawBox, {})
 * ```
 */
// export const withStyleProps = <P extends RosesStyleProps>(
export const withStyleProps = (
  componentKey: string,
  Component: React.ComponentType<RosesStyleProps>,
  options: ComposeOptions = {}
) => {
  return styled(Component)<RosesStyleProps>(
    themedComponent(componentKey),
    variantHandler(componentKey, options),
    rxHandler
  )
}
// /** Compose in the default style + variants for the key defined in your theme
//  * @example ```
//  * const theme = {
//  *   styles: {
//  *     Button: {
//  *       color: "blue"
//  *     }
//  *   }
//  * }
//  *
//  * const Button = styled('button')(ButtonProps, themed("Button"))
//  * ```
//  */
// export const themed = key => props => props.theme.styles[key]
