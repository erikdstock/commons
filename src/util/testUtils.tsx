import React from "react"
import renderer from "react-test-renderer"
import { RosesTheme } from "../Theme"

export const render = (tree: React.ReactNode) =>
  renderer.create(<RosesTheme>{tree}</RosesTheme>)
