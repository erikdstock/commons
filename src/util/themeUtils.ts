// /** Compose in the default style + variants for the key defined in your theme
//  * @example ```
//  * const theme = {
//  *   styles: {
//  *     Button: {
//  *       color: "blue"
//  *     }
//  *   }
//  * }
//  *
//  * const Button = styled('button')(ButtonProps, themed("Button"))
//  * ```
//  */
// export const themed = key => props => props.theme.styles[key]
