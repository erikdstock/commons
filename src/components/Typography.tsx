import styled from "@emotion/styled-base"
import styledCSS from "@styled-system/css"
import { boxStyle as baseStyle, withStyleProps } from "../util/styleComposition"

export const Text: RosesSC = withStyleProps("Text")
export const Heading: RosesSC = withStyleProps("Heading")
