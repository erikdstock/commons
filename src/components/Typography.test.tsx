import React from "react"
import { render } from "../util/testUtils"
import { Heading, Text } from "./Typography"

describe("Heading", () => {
  const tree = render(<Heading>Click me</Heading>).toJSON()
  it("renders correctly", () => {
    expect(tree).toMatchSnapshot()
  })
})
describe("Text", () => {
  const tree = render(<Text>Click me</Text>).toJSON()
  it("renders correctly", () => {
    expect(tree).toMatchSnapshot()
  })
})
