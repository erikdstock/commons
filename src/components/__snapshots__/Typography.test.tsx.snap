// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading renders correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  line-height: 1.125;
  font-weight: 700;
  font-family: 'Arial',system-ui,sans-serif;
  font-size: 32px;
}

<div
  className="emotion-0"
>
  Click me
</div>
`;

exports[`Text renders correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  line-height: 1.5;
  white-space: pre-wrap;
  font-family: 'Times New Roman',serif;
}

<div
  className="emotion-0"
>
  Click me
</div>
`;
